hwcomposer/wayland-hwc.cpp

```diff
@@ -460,47 +261,158 @@ struct wl_shell_surface_listener shell_surface_listener = {
+
+void surface_context::set_buffer_scale(double d_scale) {
+    assert(!viewport);
+
+    // Usually with no viewporter the scale is guaranteed to be integer
+    // When supports_cursor_viewport == false, this might not be the case
+    // thus use ceil anyway
+    int32_t scale = static_cast<int32_t>(ceil(d_scale));
+    wl_surface_set_buffer_scale(surface, scale);
+}
+
+void surface_context::set_crop(hwc_frect_t crop) {
+    assert(viewport);
+
+    float x = fmax(0, crop.left);
+    float y = fmax(0, crop.top);
+    float width = fmax(1, crop.right - crop.left);
+    float height = fmax(1, crop.bottom - crop.top);
+
+    wp_viewport_set_source(viewport,
+                           wl_fixed_from_double(x),
+                           wl_fixed_from_double(y),
+                           wl_fixed_from_double(width),
+                           wl_fixed_from_double(height));
+}
+
+void surface_context::set_display_frame(hwc_rect_t rect, double scale) {
+    assert(viewport);
+
+    int width = static_cast<int>(ceil((rect.right - rect.left) / scale));
+    int height = static_cast<int>(ceil((rect.bottom - rect.top) / scale));
+
+    wp_viewport_set_destination(viewport,
+                                std::max(1, width),
+                                std::max(1, height));
+}
+
+void window::reset_per_set_state() {
+    lastLayer = 0;
+    last_layer_buffer = nullptr;
+    if (input_region) {
+        wl_region_subtract(input_region, 0, 0, INT_MAX, INT_MAX);
+    }
+}
+
+void window::set_maximize(bool enabled) {
+    if (xdg_toplevel) {
+        if (enabled) {
+            xdg_toplevel_set_maximized(xdg_toplevel);
+        } else {
+            xdg_toplevel_unset_maximized(xdg_toplevel);
+        }
+    } else {
+        assert(shell_surface);
+        if (enabled) {
+            wl_shell_surface_set_maximized(shell_surface, display->output);
+        } else {
+            ALOGW("wl_shell_surface: does not support un-maximizing");
+        }
+    }
+}
+
+void window::set_title(const char* title) {
+    if (xdg_toplevel) {
+        xdg_toplevel_set_title(xdg_toplevel, title);
+    } else {
+        assert(shell_surface);
+        wl_shell_surface_set_title(shell_surface, title);
+    }
+}
+
+void window::set_app_id(std::string appID) {
+    this->appID = std::move(appID);
+    if (xdg_toplevel) {
+        const std::string &wayland_app_id = [&](){
+            if (this->appID != "Waydroid") {
+                return "waydroid." + this->appID;
+            } else {
+                return this->appID;
+            }
+        }();
+        xdg_toplevel_set_app_id(xdg_toplevel, wayland_app_id.c_str());
+    }
+}
+
+void window::minimize() {
+    if (xdg_toplevel) {
+        xdg_toplevel_set_minimized(xdg_toplevel);
+        wl_surface_commit(surface); // unclear if this is required
+        wl_display_flush(display->display);
+    }
+}
+
+window::~window() {
+    layers.clear();
+    if (xdg_toplevel)
+        xdg_toplevel_destroy(xdg_toplevel);
```

Comment:
This sequence generates events Wayland events in an invalid order:
> When a client wants to destroy a wl_surface, they must destroy this 'role object' before the wl_surface.

If our window has no background (e.g. because there's no `wl_subcompositor`), the `wl_surface` for this toplevel is `layer[0]` so it will be destroyed in `layers.clear()` before the toplevel role object is destroyed. Some compositors don't seem to complain, but Mutter does.